test/python/utils/test_deprecation.py

```diff
@@ -0,0 +1,464 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Tests for the functions in ``utils.deprecation``."""
+
+from textwrap import dedent
+from typing import Callable
+
+from qiskit.test import QiskitTestCase
+from qiskit.utils.deprecation import (
+    _add_deprecation_to_docstring,
+    deprecate_function,
+    deprecate_arguments,
+)
+
+
+class TestDeprecationDecorators(QiskitTestCase):
+    """Test that the decorators in ``utils.deprecation`` correctly log warnings and get added to
+    docstring."""
+
+    def test_deprecate_arguments_message(self) -> None:
+        """Test that `@deprecate_arguments` adds the correct message to the docstring."""
+
+        @deprecate_arguments(
+            {"old_arg1": "new_arg1", "old_arg2": None},
+            category=PendingDeprecationWarning,
+            since="9.99",
+        )
+        def my_func() -> None:
+            pass
+
+        self.assertEqual(
+            my_func.__doc__,
+            dedent(
+                """\
+
+                .. deprecated:: 9.99_pending
+                  TestDeprecationDecorators.test_deprecate_arguments_message.<locals>.my_func \
+keyword argument old_arg1 is deprecated and replaced with new_arg1.
```

Comment:
if you prefer, you could potentially do all these messages as a multiline regex fullmatch, or an fstring that injects `__qualname__` of the relevant object itself, so we don't need to keep the test names and error messages quite as precisely in sync. Just a suggestion, though, not a big deal either way.